### PR TITLE
Add lepton high temp mode

### LINKED ISSUE #1534: Feature Request: Support High temperature modes on FLIR Lepton 3.5 sensor
The FLIR Lepton 3.5 sensor has a high temperature output mode that supports temperature readings up to 600C.

A couple of things need to be added to support this:
- Add min and max temperature setting firmware
- Add flag for enabling high temperature mode

I have a working prototype of the changes as additions to the IOCTL interface, does it make sense to submit them as a PR?

### PR DESCRIPTION
**Pull request by linknum23**
The FLIR Lepton 3.5 sensor has a high temperature output mode that supports temperature readings up to 600C.

To support this goal the following was added:
- Add min and max temperature setting firmware
- Add flag for enabling high temperature mode

With these additions the firmware interface here seems a little flaky. Many times after setting these modes the sensor needs to be reset again for image acquisition to work. This probably shouldn't be merged until this instability is addressed. Any ideas as to why the lepton SDK is unstable here?

Fixes #1534

### CODE REVIEW

**Comment by iabdalkader** [comment]
> Any ideas as to why the lepton SDK is unstable here?

I have no idea why it becomes unstable, but maybe @kwagyeman can help, he's the expert on Lepton SDK.. 

I do have other comments on your PR though...

1) Can you please rebase your fork on our master ? The CI is failing to build your PR, it seems there's an issue with one of the submodules which was updated recently. Assuming you have a remote to our repo called upstream:

```bash
git checkout master
git fetch upstream
git rebase upstream/master
git push origin master
git checkout lepton_high_temp
git rebase master
git push --force origin lepton_high_temp
```

2) Also can you please remove the doc update for now ? It's in micropython submodule, I think it should be merged there first, then the submodule updated. You can send it in a PR to our micropython fork openmv/micropython.

3) More importantly, I would just add a 2nd argument to `IOCTL_LEPTON_SET_MEASUREMENT_MODE`  Note the ioctl function is flexible enough to support adding args, without breaking backwards compatibility. For example you can do this:
```C
case IOCTL_LEPTON_SET_MEASUREMENT_MODE:
    if (n_args >= 2) {
        bool high_temp = false;
        if (n_args > 2) {
            high_temp = mp_obj_get_int(args[2]);
       }
        error = sensor_ioctl(request, mp_obj_get_int(args[1]), &high_temp);
    }
    break;
```
Now you don't need a new command. As for `IOCTL_LEPTON_GET_MEASUREMENT_MODE` you can return a tuple, like this:
```C
case IOCTL_LEPTON_GET_MEASUREMENT_MODE: {
    int enabled, high_temp;
    error = sensor_ioctl(request, &enabled, &high_temp);
    if (error == 0) {
        ret_obj = mp_obj_new_tuple(2, (mp_obj_t []) {mp_obj_new_bool(enabled), mp_obj_new_bool(high_temp)});
    }
    break;
}
```

3) The `IOCTL_LEPTON_SET_MEASUREMENT_RANGE` the min/max temp range should be conditional on `high_temp_mode`, it should be:

```C
case IOCTL_LEPTON_SET_MEASUREMENT_RANGE: {
    float *arg_min_temp = va_arg(ap, float *);
    float *arg_max_temp = va_arg(ap, float *);
    float min_temp_range;
    float max_temp_range;
    float min_temp_range = (high_temp_mode) ? LEPTON_MIN_HIGH_TEMP : LEPTON_MIN_TEMP;
    float max_temp_range = (high_temp_mode) ? LEPTON_MAX_HIGH_TEMP : LEPTON_MAX_TEMP;
    min_temp = IM_MAX(IM_MIN(*arg_min_temp, *arg_max_temp), min_temp_range);
    max_temp = IM_MIN(IM_MAX(*arg_max_temp, *arg_min_temp), max_temp_range);
    break;
}
```

**Comment by linknum23** [comment]
Oops I was wondering what happened after I rebased onto develop. Looks like I didn't handle `ulab` properly.

Updated the branch to only have my openmv changes. Thanks so much for the interface suggestions, those make way more sense.

**Comment by iabdalkader** [comment]
Yes it looks aligned now. Let me know when you're done with all the changes and I'll rerun the workflows.

**Comment by linknum23** [comment]
Hmm something is broken on my setup I can't seem to build off the master branch even:

```
GEN openmv/src/build/lib/libtf/libtf_builtin_models.h
  File "  File "openmv/src/../tools/tflite2c.py/openmv/src/../tools/tflite2c.py", line ", line 5757

        print('  ', end='')
print('  ', end='')
                                      ^
^
SyntaxErrorSyntaxError: : invalid syntaxinvalid syntax
```

Any ideas?

**Comment by linknum23** [comment]
Also just a heads up I added the changes but won't be able to test on real hardware for a couple of days. I'll push the changes here, just realize I can't even compile them right now.

**Comment by iabdalkader** [comment]
I'm not sure, looks like `tflite2c.py` has somehow been changed.. Try `git checkout tools/tflite2c.py` ?

> Also just a heads up I added the changes but won't be able to test on real hardware for a couple of days

I can help test the change, by that I mean call the `ioctl` with the extra arg etc...

**Comment by linknum23** [comment]
Yeah I'm still having issues, even on a clean checkout. I will get back into this when I can test on real hardware.

**Comment by iabdalkader** [comment]
> Yeah I'm still having issues, even on a clean checkout. I will get back into this when I can test on real hardware.

Sure, take your time.

**Review by kwagyeman — COMMENTED** [review]
Seems fine. Does it actually work?

**Comment by linknum23** [comment]
The original changes worked but the interface was unstable. It often needed an additional reset command to start outputting images. 

New/suggested changes to the interface are untested until I get the hardware again in a couple days. Feel free to convert this to a draft pr until then.

I can also add an example script for using the camera in high temp mode.

**Comment by iabdalkader** [comment]
My suggested changes were meant to go into `py_sensor.c`, but also `lepton.c` needs to changes to accept/handle 2 args for measurement mode. Also `IOCTL_LEPTON_GET_HIGH_TEMP_MODE` `IOCTL_LEPTON_SET_HIGH_TEMP_MODE` are no longer needed/used.

File: src/omv/common/sensor.h [hunk 1]
```diff
@@ -146,6 +146,8 @@ typedef enum {
     IOCTL_LEPTON_GET_MEASUREMENT_MODE,
     IOCTL_LEPTON_SET_MEASUREMENT_RANGE,
     IOCTL_LEPTON_GET_MEASUREMENT_RANGE,
+    IOCTL_LEPTON_SET_HIGH_TEMP_MODE,
+    IOCTL_LEPTON_GET_HIGH_TEMP_MODE,
```
Comment (iabdalkader) [inline src/omv/common/sensor.h:150]:
```suggestion
```

File: src/omv/modules/py_sensor.c [hunk 2]
```diff
@@ -790,12 +790,14 @@ static mp_obj_t py_sensor_ioctl(uint n_args, const mp_obj_t *args)
         }
 
         case IOCTL_LEPTON_SET_MEASUREMENT_MODE:
+        case IOCTL_LEPTON_SET_HIGH_TEMP_MODE:
             if (n_args >= 2) {
                 error = sensor_ioctl(request, mp_obj_get_int(args[1]));
             }
             break;
```
Comment (iabdalkader) [inline src/omv/modules/py_sensor.c:797]:
```suggestion
    case IOCTL_LEPTON_SET_MEASUREMENT_MODE:
        if (n_args >= 2) {
            bool high_temp = (n_args == 2) ? false : mp_obj_get_int(args[2]);
            error = sensor_ioctl(request, mp_obj_get_int(args[1]), &high_temp);
        }
        break;
```

File: src/omv/modules/py_sensor.c [hunk 3]
```diff
@@ -790,12 +790,14 @@ static mp_obj_t py_sensor_ioctl(uint n_args, const mp_obj_t *args)
...
 
-        case IOCTL_LEPTON_GET_MEASUREMENT_MODE: {
+        case IOCTL_LEPTON_GET_MEASUREMENT_MODE:
+        case IOCTL_LEPTON_GET_HIGH_TEMP_MODE: {
             int enabled;
             error = sensor_ioctl(request, &enabled);
```
Comment (iabdalkader) [inline src/omv/modules/py_sensor.c:802]:
```suggestion
    case IOCTL_LEPTON_GET_MEASUREMENT_MODE: {
        int enabled, high_temp;
        error = sensor_ioctl(request, &enabled, &high_temp);
```

File: src/omv/modules/py_sensor.c [hunk 4]
```diff
@@ -1089,6 +1091,8 @@ STATIC const mp_map_elem_t globals_dict_table[] = {
     { MP_OBJ_NEW_QSTR(MP_QSTR_IOCTL_LEPTON_GET_AUX_TEMPERATURE),    MP_OBJ_NEW_SMALL_INT(IOCTL_LEPTON_GET_AUX_TEMPERATURE)},
     { MP_OBJ_NEW_QSTR(MP_QSTR_IOCTL_LEPTON_SET_MEASUREMENT_MODE),   MP_OBJ_NEW_SMALL_INT(IOCTL_LEPTON_SET_MEASUREMENT_MODE)},
     { MP_OBJ_NEW_QSTR(MP_QSTR_IOCTL_LEPTON_GET_MEASUREMENT_MODE),   MP_OBJ_NEW_SMALL_INT(IOCTL_LEPTON_GET_MEASUREMENT_MODE)},
+    { MP_OBJ_NEW_QSTR(MP_QSTR_IOCTL_LEPTON_SET_HIGH_TEMP_MODE),     MP_OBJ_NEW_SMALL_INT(IOCTL_LEPTON_SET_HIGH_TEMP_MODE)},
+    { MP_OBJ_NEW_QSTR(MP_QSTR_IOCTL_LEPTON_GET_HIGH_TEMP_MODE),     MP_OBJ_NEW_SMALL_INT(IOCTL_LEPTON_GET_HIGH_TEMP_MODE)},
```
Comment (iabdalkader) [inline src/omv/modules/py_sensor.c:1095]:
```suggestion
```

File: src/omv/sensors/lepton.c [hunk 5]
```diff
@@ -279,24 +284,27 @@ static int ioctl(sensor_t *sensor, int request, va_list ap)
...
         }
-        case IOCTL_LEPTON_SET_MEASUREMENT_MODE: {
-            int enabled = va_arg(ap, int);
-            if (measurement_mode != enabled) {
-                measurement_mode = enabled;
-                ret = lepton_reset(sensor, measurement_mode);
+        case IOCTL_LEPTON_SET_MEASUREMENT_MODE:
+            if (n_args >= 2) {
+                bool high_temp = (n_args == 2) ? false : mp_obj_get_int(args[2]);
+                error = sensor_ioctl(request, mp_obj_get_int(args[1]), &high_temp);
```
Comment (iabdalkader) [inline src/omv/sensors/lepton.c:290]:
```suggestion
        case IOCTL_LEPTON_SET_MEASUREMENT_MODE:
            int enabled = va_arg(ap, int);
            int high_temp = va_arg(ap, int);
            if (measurement_mode != enabled) {
                measurement_mode = enabled;
                high_temp_mode = high_temp;
                ret = lepton_reset(sensor, measurement_mode, high_temp_mode);
```

**Review by iabdalkader — CHANGES_REQUESTED** [review]

**Comment by iabdalkader** [comment]
@linknum23  I left a full review, hopefully it clears out any confusion.

File: src/omv/sensors/lepton.c [hunk 6]
```diff
@@ -40,15 +40,20 @@
 #define VOSPI_FIRST_PACKET      (0)
 #define VOSPI_FIRST_SEGMENT     (1)
 #define LEPTON_TIMEOUT          (1000)
+// Temperatures in Celsius
 #define DEFAULT_MIN_TEMP        (-17.7778f)
 #define DEFAULT_MAX_TEMP        (37.7778f)
+#define LEPTON_MIN_TEMP            (-10.0f)
+#define LEPTON_MAX_TEMP            (600.0f)
+
```
Comment (iabdalkader) [inline src/omv/sensors/lepton.c:48]:
```suggestion
#define LEPTON_MIN_TEMP            (-10.0f)
#define LEPTON_MAX_TEMP            (140.0f)
#define LEPTON_MIN_HIGH_TEMP       (-10.0f)
#define LEPTON_MAX_HIGH_TEMP       (600.0f)
```

**Review by iabdalkader — CHANGES_REQUESTED** [review]

File: src/omv/sensors/lepton.c [hunk 7]
```diff
@@ -40,15 +40,21 @@
 #define VOSPI_FIRST_PACKET      (0)
 #define VOSPI_FIRST_SEGMENT     (1)
 #define LEPTON_TIMEOUT          (1000)
+// Temperatures in Celsius
 #define DEFAULT_MIN_TEMP        (-17.7778f)
```
Comment (linknum23) [inline src/omv/sensors/lepton.c:44]: -17.77 is outside the sensors range, any chance I can make this -10.0 or am I missing something here?

**Comment by linknum23** [comment]
> Sure, take your time.

Getting back into this and I had time to figure out my issues. Mostly just missed this line:
`git -C src/micropython/ submodule update --init --depth=1`

File: src/omv/sensors/lepton.c (hunk 7 above)
Comment (iabdalkader) [reply src/omv/sensors/lepton.c:44]: I'm not sure, CC @kwagyeman

**Comment by iabdalkader** [comment]
Hi, there were still some issues with the changes, some formatting issues, and issues with passing args etc... all minor fixes, so I just reworked the PR here, keeping your commits: https://github.com/openmv/openmv/pull/1542

I also tested it, and I don't see any instability. In fact it seems more sensitive, or scales better, with high temp is enabled. With high temp disabled it looks like all temps are flattened.

File: src/omv/sensors/lepton.c (hunk 7 above)
Comment (kwagyeman) [reply src/omv/sensors/lepton.c:44]: It's 0 in Fahrenheit. Please make it -10.

**Comment by iabdalkader** [comment]
Maybe this should be the default mode.

**Comment by iabdalkader** [comment]
It's now merged in master, but I'm leaving this open for discussion,

File: src/omv/sensors/lepton.c (hunk 7 above)
Comment (iabdalkader) [reply src/omv/sensors/lepton.c:44]: What about the default max ? Could it need fixing too ? The measurement mode looks much better with the higher range, see my comment below.

**Comment by iabdalkader** [comment]
@linknum23 Hi, I already squashed and merged your commits with some fixes and cleanup here: https://github.com/openmv/openmv/commit/6b2cd1008861a1d763ef0568938f072c87f990a0

If you have more fixes please send them in new PRs.

**Comment by linknum23** [comment]
Oops didn't notice this. Had my head down...
Thanks for all of the help on this!

I moved some things around to make sure radiometry mode was set correctly all of the time. Let me play with the changes to master and see how they work.

**Pull request closed without merging**